docsp-32163/32167 - quick start and introduction (#1)
mongodb/docs-entity-framework: source/quick-start.txt

.. _entity-framework-quickstart:

===========
Quick Start
===========

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

.. include:: /includes/quick-start/overview.rst

Create a MongoDB Cluster
------------------------

> Set Up a Free Tier Cluster in Atlas

Author: Changed to use procedures. Worth noting that several of our Quick Starts (e.g., Java, Go) use stacked titles.

The Quick Start opening paragraph says "MongoDB Atlas." Does that make this heading sufficiently clear?

> Set Up a Free Tier Cluster in Atlas

Reviewer: > The Quick Start opening paragraph says "MongoDB Atlas." Does that make this heading sufficiently clear?

Sorry if I had missed that!

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

To set up your Atlas Free Tier Cluster required for this guide, complete the guide on 
:guides:`MongoDB Atlas Setup </atlas/account>`.

After completing the steps in the Atlas guide, you have a new MongoDB
cluster deployed in Atlas, a new database user, and

> :atlas:`sample datasets loaded </sample-data/>` into your cluster. You also have

Reviewer: Suggestion:
I think the link could be more descriptive in terms of what it's linking, or be removed since it references part of the content contained within the previous link. Here are some[ link and cross-reference examples](https://www.mongodb.com/docs/meta/style-guide/style/links-and-cross-references/#link-and-cross-reference-examples) in the style guide.

a connection string similar to the following in your copy buffer:

.. code-block:: bash

>    "mongodb+srv://<username>:<password>@cluster0.abc.mongodb.net/?retryWrites=true&w=majority"

Reviewer: Suggestion:

Since this isn't intended to be copied (as the user would overwrite their actual connection string), I think it would be helpful to set this to be "copyable: false".

Set Your Connection String
~~~~~~~~~~~~~~~~~~~~~~~~~~

> Run the following code at the command prompt to save your MongoDB connection string to an 

Author: What if we just say "in your shell" [like we do in Node?](https://www.mongodb.com/docs/drivers/node/current/quick-start/download-and-install/)

environment variable. This method is safer than including your credentials in your source 
code. 

.. code-block:: bash

   export MONGODB_URI='<your MongoDB URI>'

.. important::

   Make sure to replace the ``<username>`` and ``<password>`` sections of the connection 

>    string with the username and password of your Atlas user.

Reviewer: Never mind, I was looking at the guide again and they will need to replace username and password. However, I think it would be helpful to mention recording your database user's credentials somewhere, similar to [these instructions](https://www.mongodb.com/docs/drivers/node/current/quick-start/create-a-connection-string/#update-the-placeholders).

Set Up Your Project
-------------------

Create the Project
~~~~~~~~~~~~~~~~~~

Create a new directory and initialize your project with the ``dotnet new`` command,
as follows:

.. code-block:: shell

   mkdir entity-quickstart
   cd entity-quickstart
   dotnet new console

.. _entity-framework-add-mongodb-dependency:

> Add MongoDB as a Dependency

Reviewer: Question:

Should this be the provider, not MongoDB?

Author: Yes, that's more accurate.

~~~~~~~~~~~~~~~~~~~~~~~~~~~

Use the ``dotnet add`` command to add the {+provider-short+} to your project as a
dependency.

.. code-block:: shell

>    dotnet add package [todo: insert package here] 

Reviewer: Suggestion:
Use a source constant as a placeholder instead of "todo" since it will likely be used in more than one place.

Author: Good idea!

Query Your MongoDB Cluster from Your Application
------------------------------------------------

In this step, you'll use the {+provider-short+} 
to connect to your MongoDB cluster and run a LINQ query on the sample data.

Open the file named ``Program.cs`` in the base directory of your project. Copy the
following sample code into ``Program.cs``: 

.. literalinclude:: /includes/quick-start/Program.cs
   :language: csharp
   :dedent:

> This sample code runs a query against your sample dataset in MongoDB Atlas. Run it

Reviewer: Suggestion:

I think "against" is unnecessary; I think "query your sample dataset" has the same intended message. I think it's also ok to use "Atlas" since it was used that way since the first step of the first heading ("Set Up a Free Tier Cluster in Atlas").

from your command line by using the following command:

.. code-block:: bash

   dotnet run entity-quickstart.csproj

.. include:: /includes/quick-start/query-output.rst

.. tip::

   If your output is empty, ensure you have loaded the
   :atlas:`sample datasets </sample-data/>` into your cluster.

After completing this step, you should have a working application that uses
the {+provider-short+} to connect to your MongoDB cluster, run a query on the

> sample data, and print out the result.

Reviewer: Question:
Is the C#/.NET driver completing most of these actions or is it the provider? I thought the provider was just an ORM which means it is just serializing the data into and from classes. I think it would be inaccurate to describe it as connecting to your cluster, running the query, and printing the result.

Author: You're correct; the provider uses the driver to connect/load/save. Updated.

.. _entity-framework-quickstart:

===========
Quick Start
===========

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

.. include:: /includes/quick-start/overview.rst

Create a MongoDB Cluster
------------------------

.. procedure::
   :style: connected

   .. step:: Set Up a Free Tier Cluster in Atlas

>       To set up your Atlas free cluster required for this Quick Start, complete the guidei

Reviewer: Suggestion:

Typo fix

```suggestion
      To set up your Atlas free cluster required for this Quick Start, complete the guide
```

      on :guides:`MongoDB Atlas Setup </atlas/account>`.

      After completing the steps in the Atlas guide, you have a new MongoDB
      cluster deployed in Atlas, a new database user, and
      sample datasets loaded into your cluster. You also have
      a connection string similar to the following in your copy buffer:

      .. code-block:: bash
         :copyable: false

         "mongodb+srv://<username>:<password>@cluster0.abc.mongodb.net/?retryWrites=true&w=majority"

   .. step:: Update the Placeholders

     Paste the connection string in your copy buffer into a file in your preferred text
     editor. Replace the ``<username>`` and ``<password>`` placeholders with
     your database user's username and password.

     Save this file to a safe location for use in the next step.

>    .. step:: Set Your Connection String

Reviewer: Suggestion:
I think the connection between copying your connection string to an environment variable and "set your connection string" seems more distant than necessary. I think "Add your Connection String to an Environment Variable" would more accurately describe this step.

      Run the following code in your shell to save the MongoDB connection string
      in your copy buffer from the previous step to an 

>       environment variable. This method is safer than including your credentials in your source 

Reviewer: Suggestion:
I think when using subjective comparisons such as "safer", it would help the user understand why by clarifying the reason why. E.g.

"This method of storing your connection string in an environment variable keeps your credentials separate from your source code. This separation makes it less likely to expose your credentials when sharing your code."

      code. 

   .. code-block:: bash

      export MONGODB_URI='<your connection string>'

   .. important::

      Make sure to replace the ``<username>`` and ``<password>`` sections of the connection 

>       string with the username and password of your Atlas user.

Reviewer: Issue:

I think "Atlas user" is incorrect because that I could confuse that with the email/password combination that I used to create my Atlas account.

Suggestion:
I think this should read "database user" similar to the other instance to prevent confusion.

```suggestion
      string with the username and password of your database user.
```

Set Up Your Project
-------------------

.. procedure::
   :style: connected

   .. step:: Create the Project

>       Create a new directory and initialize your project with the ``dotnet new`` command,

Reviewer: Issue:

Style guide on "[using/with](https://www.mongodb.com/docs/meta/style-guide/terminology/alphabetical-terms/#u)":

> Use by using, or recast the sentence. Don't use using (alone) or with. By using is easier for worldwide audiences to understand.

Suggestion:

```suggestion
      Create a new directory and initialize your project by using the ``dotnet new`` command,
```

      as follows:

      .. code-block:: shell

         mkdir entity-quickstart
         cd entity-quickstart
         dotnet new console

   .. _entity-framework-add-mongodb-dependency:

   .. step:: Add the {+provider-long+} as a Dependency

      Use the ``dotnet add`` command to add the {+provider-short+} to your project as a
      dependency.

      .. code-block:: shell

         dotnet add package {+package+} 

Query Your MongoDB Cluster from Your Application
------------------------------------------------

.. procedure::
   :style: connected

>    .. step:: Create the Code File

Reviewer: Suggestion:

I think this task includes more than creating a file since it includes adding sample code to it. Perhaps a more descriptive name could be "Add the Sample Code" or something similar.

      Open the file named ``Program.cs`` in the base directory of your project. Copy the
      following sample code into ``Program.cs``: 

      .. literalinclude:: /includes/quick-start/Program.cs
         :language: csharp
         :dedent:

>    .. step:: Run an Atlas Query

Reviewer: Suggestion:

I think this is not a query specific to Atlas, but rather MongoDB. I think that info might be too detailed and could be omitted. This title might resemble the reference in the last sentence of this section more closely:

```suggestion
   .. step:: Query the Sample Data
```

      Run the following command in your shell. It should print the plot of the movie
      "Back to the Future" from the sample dataset:

      .. io-code-block::
         :copyable: true

         .. input::
            :language: none
            
            dotnet run entity-quickstart.csproj

         .. output::
            :language: none
            :visible: false

            A young man is accidentally sent 30 years into the past in a time-traveling
            DeLorean invented by his friend, Dr. Emmett Brown, and must make sure his
            high-school-age parents unite in order to save his own existence.

      .. tip::

         If your output is empty, ensure you have loaded the
         :atlas:`sample datasets </sample-data/>` into your cluster.

After completing these steps, you should have a working {+framework+} application that
connects to your MongoDB cluster, runs a query on the
sample data, and prints out the result.

Next Steps
----------

> Learn how to use the {+provider-short+} to perform common operations in Quick Reference.

Reviewer: Suggestion:

I am not sure whether this was intentionally left out, but I think this is missing a link.
